### PR TITLE
apply TypeMap into matching/notMatching/innerJoinWith...

### PR DESCRIPTION
**Pull request by saeideng**
apply TypeMap into matching/notMatching/innerJoinWith/leftJoinWith
i just tested for innerJoinWith() and i am not sure these changes work well for all case

follow: #7506

### CODE REVIEW

**Comment by saeideng** [comment]
if you agree with this commit maybe i can add some tests

**Comment by markstory** [comment]
Anyway we could get tests to help prevent any future regressions?

**Comment by saeideng** [comment]
I will can add some tests, but i need know you agree with these codes/add typemap

**Comment by markstory** [comment]
@saeideng I'm not against joined relations having typemaps set.

**Comment by saeideng** [comment]
but innerJoinWith() (for example) returned invalid query

**Comment by saeideng** [comment]
$query->innerJoinWith(.. )
 ```
->Where([
            'Table.active'=>false,
            'Table.created >='=>new \Cake\I18n\Time('24 hours ago'),
            ])
//returned query:
WHERE 
  (
    Table.active = '' 
    AND Table.created >= '9/25/16, 8:14 PM'
  ) 
```
but i can use 
`'Table.created >='=>(new \Cake\I18n\Time('24 hours ago'))->format('Y-m-d H:i:s'),` for fix it

**Comment by markstory** [comment]
@saeideng These changes would fix the error your having though wouldn't they? In my previous comment I was trying to say that these changes make sense to me :+1:

**Comment by lorenzo** [comment]
@saeideng do you think you can add tests to this?

**Comment by saeideng** [comment]
yes,but tomorrow

**Comment by saeideng** [comment]
added tests

File: tests/TestCase/ORM/QueryRegressionTest.php [hunk 1]
```diff
@@ -1206,6 +1253,36 @@ public function testComplexNestedTypesInJoinedWhere()
...
+
+        $query = $table->find()
+            ->innerJoinWith('Comments.Articles.Authors')
+            ->where([
+                'Authors.created >' => new \DateTime('2007-03-17 01:16:00')
+            ]);
+
+        $result = $query->first();
+        $this->assertNotEmpty($result);
+        $this->assertInstanceOf('Cake\I18n\Time', $result->updated);
```
Comment (markstory) [inline tests/TestCase/ORM/QueryRegressionTest.php]: Couldn't these assertions be made on the results created on line 1263?

File: tests/TestCase/ORM/QueryRegressionTest.php [hunk 2]
```diff
@@ -1206,6 +1253,36 @@ public function testComplexNestedTypesInJoinedWhere()
...
+
+        $query = $table->find()
+            ->leftJoinWith('Comments.Articles.Authors')
+            ->where([
+                'Authors.created >' => new \DateTime('2007-03-17 01:16:00')
+            ]);
+
+        $result = $query->first();
+        $this->assertNotEmpty($result);
+        $this->assertInstanceOf('Cake\I18n\Time', $result->updated);
```
Comment (markstory) [inline tests/TestCase/ORM/QueryRegressionTest.php]: This looks like a redundant test, but I might be missing something.

File: tests/TestCase/ORM/QueryRegressionTest.php (hunk 1 above)
Comment (saeideng) [inline tests/TestCase/ORM/QueryRegressionTest.php]: 1263 returned matching result
But this,has result of innerJoinWith()

File: tests/TestCase/ORM/QueryRegressionTest.php (hunk 2 above)
Comment (saeideng) [inline tests/TestCase/ORM/QueryRegressionTest.php]: i fact,we no neet to assertInstanceOf() in these tests totally :)
But i am not sure

Comment (markstory) [inline tests/TestCase/ORM/QueryRegressionTest.php]: I meant that the query and results would be the same as a test earlier in the method. The instanceOf check is useful to ensure the correct object was hydrated.

File: tests/TestCase/ORM/QueryRegressionTest.php (hunk 1 above)
Comment (markstory) [inline tests/TestCase/ORM/QueryRegressionTest.php]: My mistake. Perhaps separate test methods would have made that clearer.

**Pull request merged**